File: src/main/java/de/take_weiland/mods/commons/util/Sides.java
package de.take_weiland.mods.commons.util;

import net.minecraft.entity.Entity;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.relauncher.Side;

/**
 * <p>Helper methods for distinguishing between client and server.</p>
 */
public final class Sides {

    /**
     * <p>Get the logical side (based on {@link World#isRemote}) of the given World.</p>
     * <p>This method is best used with a static import like so:<code><pre>
     *     if (sideOf(world).isClient()) {
     *         // do client-only things
     *     }
     * </pre></code></p>
     *
     * @param world the World
     * @return the logical side
     */
    public static Side sideOf(World world) {
        return world.isRemote ? Side.CLIENT : Side.SERVER;
    }

    /**
     * <p>Get the logical side (based on {@link World#isRemote}) of the given Entity.</p>
     * <p>This method is best used with a static import like so:<code><pre>
     *     if (sideOf(entity).isClient()) {
     *         // do client-only things
     *     }
     * </pre></code></p>
     *
     * @param entity the Entity
     * @return the logical side
     */
    public static Side sideOf(Entity entity) {
        return sideOf(entity.world);
    }

    /**
     * <p>Get the logical side (based on {@link World#isRemote}) of the given TileEntity.</p>
     * <p>This method is best used with a static import like so:<code><pre>
     *     if (sideOf(tileEntity).isClient()) {
     *         // do client-only things
     *     }
     * </pre></code></p>
     *
     * @param tileEntity the TileEntity
     * @return the logical side
     */
    public static Side sideOf(TileEntity tileEntity) {
        return sideOf(tileEntity.getWorld());
    }

    /**
     * <p>Get the environment we are running on. {@code Side.SERVER} for a dedicated server or
     * {@code Side.CLIENT} for the client.</p>
     *
     * @return the environment
     */
    public static Side environment() {
        return FMLCommonHandler.instance().getSide();
    }

    private Sides() {
    }
}

File: src/main/kotlin/de/take_weiland/mods/commons/util/sides.kt
package de.take_weiland.mods.commons.util

import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.inventory.Container
import net.minecraft.server.MinecraftServer
import net.minecraft.tileentity.TileEntity
import net.minecraft.world.World
import net.minecraft.world.WorldServer
import net.minecraftforge.fml.common.FMLCommonHandler
import net.minecraftforge.fml.relauncher.Side

/**
 * @author diesieben07
 */

val World.side inline get() = if (isRemote) Side.CLIENT else Side.SERVER
val World.isServer inline get() = !isRemote
val World.isClient inline get() = isRemote

inline fun World.ifServer(body: WorldServer.() -> Unit) {
    if (isServer) (this as WorldServer).body()
}

inline fun World.requireServer(): WorldServer = if (isServer) this as WorldServer else throw IllegalStateException("WorldServer expected")

val TileEntity.side inline get() = world.side
val TileEntity.isServer inline get() = world.isServer
val TileEntity.isClient inline get() = world.isClient

val Entity.side inline get() = world.side
val Entity.isServer inline get() = world.isServer
val Entity.isClient inline get() = world.isClient

val Container.side get(): Side = if (isClient) Side.CLIENT else Side.SERVER
val Container.isServer get() = listeners.any { it is EntityPlayerMP }
val Container.isClient inline get() = !isServer

val serverInstance : MinecraftServer get() = FMLCommonHandler.instance().minecraftServerInstance